Quantization oneDNN backend only support VNNI CPU

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103653

**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

leslie-fang-intel committed:
Quantization oneDNN backend only support VNNI CPU

[ghstack-poisoned]

jgong5 commented on docs/source/quantization.rst:
```diff
@@ -1009,6 +1009,9 @@ if ``dtype`` is ``torch.qint8``, make sure to set a custom ``quant_min`` to be `
 you call the `torch.ao.quantization.get_default_qconfig(backend)` or `torch.ao.quantization.get_default_qat_qconfig(backend)` function to get the default ``qconfig`` for
 ``x86`` or ``qnnpack`` backend
 
+2. If you are using the ``onednn`` backend, we will use 8 bits for activation in the default qconfig mapping ``torch.ao.quantization.get_default_qconfig_mapping('onednn')``
+and qconfig ``torch.ao.quantization.get_default_qconfig('onednn')``. It means onednn quantization backend only support CPU with VNNI instruction.
```
```suggestion
2. If you are using the ``onednn`` backend, we will use 8 bits for activation in the default qconfig mapping ``torch.ao.quantization.get_default_qconfig_mapping('onednn')``
and qconfig ``torch.ao.quantization.get_default_qconfig('onednn')``. The ``onednn`` quantization backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
```

leslie-fang-intel committed:
Update docs/source/quantization.rst

Co-authored-by: Jiong Gong <jiong.gong@intel.com>

leslie-fang-intel commented:
Updated. Please help to take a look again.

malfet commented on docs/source/quantization.rst:
```diff
@@ -1009,6 +1009,9 @@ if ``dtype`` is ``torch.qint8``, make sure to set a custom ``quant_min`` to be `
 you call the `torch.ao.quantization.get_default_qconfig(backend)` or `torch.ao.quantization.get_default_qat_qconfig(backend)` function to get the default ``qconfig`` for
 ``x86`` or ``qnnpack`` backend
 
+2. If you are using the ``onednn`` backend, we will use 8 bits for activation in the default qconfig mapping ``torch.ao.quantization.get_default_qconfig_mapping('onednn')``
```
English is not my native, but shouldn't documentation better be in passive tense
```suggestion
2. If ``onednn`` backend is selected, 8 bits for activation in the default qconfig mapping ``torch.ao.quantization.get_default_qconfig_mapping('onednn')`` will be used.
```

leslie-fang-intel replied:
Thanks for the suggestions. Update it.

leslie-fang-intel committed:
Update docs/source/quantization.rst

Co-authored-by: Nikita Shulga <nshulga@meta.com>

andravin commented on docs/source/quantization.rst:
```diff
@@ -1009,6 +1009,9 @@ if ``dtype`` is ``torch.qint8``, make sure to set a custom ``quant_min`` to be `
 ``x86`` or ``qnnpack`` backend
 
+2. If ``onednn`` backend is selected, 8 bits for activation in the default qconfig mapping ``torch.ao.quantization.get_default_qconfig_mapping('onednn')`` will be used.
+and qconfig ``torch.ao.quantization.get_default_qconfig('onednn')``. The ``onednn`` quantization backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
```
"only support VNNI CPU" is much stricter than "recommended to be used on CPUs with Vector Neural Network Instruction support."

If the `onednn` quantization backend only supports processors with VNNI, then the documentation should say that. Also, the code should warn the user when the `onednn` backend is used on a CPU without VNNI or throw an exception.

leslie-fang-intel replied:
Thanks for the comment. 
I am working on to add the warning, will keep you updated.

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"




[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"




[ghstack-poisoned]

leslie-fang-intel commented:
Hi @andravin @jgong5, the warning messages are added. Please help to take a look again.

andravin commented:
@leslie-fang-intel The new warning message would have helped me, thanks. The warning message has minor grammar issues; proofreading with a service like Grammarly might help.

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


[ghstack-poisoned]

leslie-fang-intel commented:
> @leslie-fang-intel The new warning message would have helped me, thanks. The warning message has minor grammar issues; proofreading with a service like Grammarly might help.

Thanks for the point out and changed.

jgong5 commented on aten/src/ATen/native/CPUUtils.cpp:
Would `aten/src/ATen/cpu` a better folder to hold this?

leslie-fang-intel replied:
Thanks for the suggestion and move to `aten/src/ATen/cpu` folder.

jgong5 commented on docs/source/quantization.rst:
```diff
@@ -1009,6 +1009,11 @@ if ``dtype`` is ``torch.qint8``, make sure to set a custom ``quant_min`` to be `
 you call the `torch.ao.quantization.get_default_qconfig(backend)` or `torch.ao.quantization.get_default_qat_qconfig(backend)` function to get the default ``qconfig`` for
 ``x86`` or ``qnnpack`` backend
 
+2. If ``onednn`` backend is selected, 8 bits for activation will be used in the default qconfig mapping ``torch.ao.quantization.get_default_qconfig_mapping('onednn')``
+and default qconfig ``torch.ao.quantization.get_default_qconfig('onednn')``. It is recommended to be used on CPUs with Vector Neural Network Instruction
+support. Otherwise, setting ``reduce_range`` to True of the activation's observer to get better accuracy on CPUs without Vector Neural Network Instruction
+support.
```
```suggestion
2. If ``onednn`` backend is selected, 8 bits for activation will be used in the default qconfig mapping ``torch.ao.quantization.get_default_qconfig_mapping('onednn')``
and default qconfig ``torch.ao.quantization.get_default_qconfig('onednn')``. It is recommended to be used on CPUs with Vector Neural Network Instruction (VNNI)
support. Otherwise, setting ``reduce_range`` to True of the activation's observer to get better accuracy on CPUs without VNNI
support.
```

leslie-fang-intel replied:
Thanks and changed.

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented on torch/csrc/autograd/init.cpp:
```diff
@@ -837,6 +848,7 @@ static PyMethodDef methods[] = { // NOLINT
     {"get_autocast_cpu_dtype", get_autocast_cpu_dtype, METH_NOARGS, nullptr},
     {"set_autocast_gpu_dtype", set_autocast_gpu_dtype, METH_O, nullptr},
     {"get_autocast_gpu_dtype", get_autocast_gpu_dtype, METH_NOARGS, nullptr},
+    {"_is_cpu_support_vnni", is_cpu_support_vnni, METH_NOARGS, nullptr},
```
Sure, move the binding into `torch/csrc/cpu` and expose it under `torch.cpu`.

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/cpu/CPUUtils.cpp:
Since it is already under `cpu` folder, maybe name it `utils.cpp` is good enough?

leslie-fang-intel replied:
Thanks and rename it.

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Quantization oneDNN backend only support VNNI CPU"


**Summary**

- Update the quantization document that default qconfig with oneDNN backend is recommended to be used on CPUs with Vector Neural Network Instruction support.
- Add the warning message when user uses default qconfig with oneDNN backend on CPU without Vector Neural Network Instruction support.


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
